neuralop/training/trainer.py

```diff
@@ -130,173 +93,163 @@ def train(
                     self.log_epoch(
                         epoch=epoch,
                         time=epoch_train_time,
                         avg_loss=avg_loss,
                         train_err=train_err,
                         avg_lasso_loss=avg_lasso_loss,
-                        eval_metrics=all_errors,
+                        eval_metrics=all_metrics,
                         lr=lr
                     )
+                
+                # save checkpoint if conditions are met
+                if save_best is not None:
+                    if all_metrics[save_best] < best_metric_value:
+                        best_metric_value = all_metrics[save_best]
+                        self.checkpoint(save_dir)
+
+            # save checkpoint if save_every and save_best is not set
+            if self.save_every is not None:
+                if epoch % self.save_every == 0:
+                    print(f"Saving state to {save_dir}")
+                    self.checkpoint(save_dir)
+
+        return all_metrics
+
+    def train_one_epoch(self, epoch, train_loader, training_loss):
+        """train_one_epoch trains self.model on train_loader
+        for one epoch and returns training metrics
+
+        Parameters
+        ----------
+        epoch : int
+            epoch number
+        train_loader : torch.utils.data.DataLoader
+            data loader of train examples
+        test_loaders : dict
+            dict of test torch.utils.data.DataLoader objects
+
+        Returns
+        -------
+        all_errors
+            dict of all eval metrics for the last epoch
+        """
+        self.on_epoch_start(epoch)
+        avg_loss = 0
+        avg_lasso_loss = 0
+        self.model.train()
+        if self.data_processor:
+            self.data_processor.train()
+        t1 = default_timer()
+        train_err = 0.0
+        
+        # track number of training examples in batch
+        self.n_samples = 0
+
+        for idx, sample in enumerate(train_loader):
+            
+            loss = self.train_one_batch(idx, sample, training_loss)
+            train_err += loss.item()
 
-                if self.callbacks:
-                    self.callbacks.on_val_end()
+            loss.backward()
+            #del out
+            self.optimizer.step()
 
-            if self.callbacks:
-                self.callbacks.on_epoch_end(
-                    epoch=epoch, train_err=train_err, avg_loss=avg_loss
-                )
+            with torch.no_grad():
+                avg_loss += loss.item()
+                if self.regularizer:
+                    avg_lasso_loss += self.regularizer.loss
 
-        return all_errors
+        if isinstance(self.scheduler, torch.optim.lr_scheduler.ReduceLROnPlateau):
+            self.scheduler.step(train_err)
+        else:
+            self.scheduler.step()
+
+        epoch_train_time = default_timer() - t1
+
+        train_err /= len(train_loader)
+        avg_loss /= self.n_samples
```

Comment:
Separately, (not in the scope of this PR), -perhaps we should remove this averaging and leave it to the loss to normalize -or not. 

Either way, len(train_loader) would be the batch size here, *not* the length of the dataset, we should use the accumulated n_samples

Reply:
We should still divide by number of batches though, right?